[LoopVectorize] Generate wide active lane masks

kmclaughlin-arm opened the pull request:
This patch adds a new flag (-enable-wide-lane-mask) which allows
LoopVectorize to generate wider-than-VF active lane masks when it
is safe to do so (i.e. the mask is used for data and control flow).

The transform in extractFromWideActiveLaneMask creates vector
extracts from the first active lane mask in the header & loop body,
modifying the active lane mask phi operands to use the extracts.

An additional operand is passed to the ActiveLaneMask instruction,
the value of which is used as a multiplier of VF when generating the mask.
By default this is 1, and is updated to UF by extractFromWideActiveLaneMask.

The motivation for this change is to improve interleaved loops when
SVE2.1 is available, where we can make use of the whilelo instruction
which returns a predicate pair.

This is based on a PR that was created by @momchil-velikov (#81140)
and contains tests which were added there.

kmclaughlin-arm committed:
[LoopVectorize] Generate wide active lane masks

This patch adds a new flag (-enable-wide-lane-mask) which allows
LoopVectorize to generate wider-than-VF active lane masks when it
is safe to do so (i.e. using tail folding without runtime checks).

The transform in extractFromWideActiveLaneMask creates vector
extracts from the first active lane mask in the header & loop body,
modifying the active lane mask phi operands to use the extracts.

An additional operand is passed to the ActiveLaneMask instruction, the
value of which is used as a multiplier of VF when generating the mask. By
default this is 1, and is updated to UF by extractFromWideActiveLaneMask.

The motivation for this change is to improve interleaved loops when
SVE2.1 is available, where we can make use of the whilelo instruction
which returns a predicate pair.

This is based on a PR that was created by @momchil-velikov (#81140)
and contains tests which were added there.

kmclaughlin-arm committed:
- Replaced --passes="default<O3>" with --passes=loop-vectorize
- Used the --filter-out-after option when regenerating test CHECK lines
- Removed CodeGen test as similar tests exist in get-active-lane-mask-extract.ll

kmclaughlin-arm commented on llvm/test/Transforms/LoopVectorize/AArch64/fixed-wide-lane-mask.ll:
```diff
@@ -0,0 +1,129 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
+; RUN: opt -S --passes="default<O3>" -scalable-vectorization=off -prefer-predicate-over-epilogue=predicate-dont-vectorize -enable-wide-lane-mask \
```
I've changed the tests to only use `-passes=loop-vectorize` & regenerated them with the `--filter-out-after` flag to reduce the size instead.
In the updated tests, the extractelement always uses the first `@llvm.vector.extract` so I will investigate the possible missing instcombine separately.

kmclaughlin-arm committed:
- Removed VPInstruction::ExtractSubvector & use VPWidenIntrinsicRecipe
- Removed UnrollPart from VPActiveLaneMaskPHIRecipe & order based on phi backedge values
- Include multiplier in VPInstruction::computeCost for ActiveLaneMask
- Call useActiveLaneMask from optimizeForVFAndUF instead of simplifyBranchConditionForVFAndUF
- Always enable wide active lane masks if flag is passed, not only if DataAndControlFlowWithoutRuntimeCheck
- Added an assert that the incoming values to the first Phi are ActiveLaneMask instructions
- Change -force-vector-interleave=0 to -force-vector-interleave=1 & removed tests for UF2
- Moved EnableWideActiveLaneMask to VPlanTransforms.cpp

kmclaughlin-arm committed:
- Rebase & address review comments in VPlanTransforms.cpp
 - Update tests after rebasing

david-arm reviewed (approved):
LGTM with nit addressed! I think we can probably clean up the dead active lane masks in a follow-up patch, since the new flag is off by default anyway.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1446,6 +1448,92 @@ static bool isConditionTrueViaVFAndUF(VPValue *Cond, VPlan &Plan,
+
+  using namespace llvm::VPlanPatternMatch;
+  if (!EnableWideActiveLaneMask || !VF.isVector() || UF == 1 ||
+      !match(Term, m_BranchOnCond(m_Not(m_ActiveLaneMask(
```
The description of the patch says
> when it is safe to do so (i.e. using tail folding without runtime checks).

Is checking for ActiveLaneMask here enough to ensure that the unsafe cases are rejected?

kmclaughlin-arm committed:
- Move definition of EnableWideActiveLaneMask into VPlanTransforms
- Move some checks to the top of useWideActiveLaneMask
- Add documentation to functions & rename useWideActiveLaneMask

kmclaughlin-arm committed:
- Address missed comments in sve-wide-lane-mask.ll

kmclaughlin-arm committed:
- Fix the comment describing ActiveLaneMask in VPlan.h

kmclaughlin-arm committed:
- Rebase and run update_test_checks.py on sve-wide-lane-mask.ll

kmclaughlin-arm committed:
- Replace fallthrough with break to fix test failures after rebasing

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -4214,9 +4214,15 @@ VectorizationFactor LoopVectorizationPlanner::selectVectorizationFactor() {
                 }
               }
             }
-            [[fallthrough]];
```
Hmm, I don't think this is right because previously when falling through it was hitting the VPInstruction::ExplicitVectorLength case below and adding on the cost of the instruction, i.e. `C += VPI->cost(VF, CostCtx)`.

If you change this to

```
  C += VPI->cost(VF, CostCtx);
  break;
```

then it will be the same as before.

kmclaughlin-arm committed:
- Add VPI->cost back into Select case in selectVectorizationFactor

david-arm reviewed (approved):
LGTM. Seems like you've addressed all the comments now - thank you!

fhahn reviewed (commented):
Thanks for all the updates, LG!
